enhancements/security/manual-cloud-credentials.md

```diff
@@ -0,0 +1,170 @@
+
+Cloud credential operator reconcile code will be refactored to be a little more friendly to use as a vendored library, for use in the CLI.
+
+Note that when we say CCO is disabled, it is still running, just not minting credentials, and does not have access to the powerful "root" credentials for the cloud provider account.
+
+### Risks and Mitigations
+
+
+## Design Details
+
+Core of this proposal will be CLI tooling which vendors code from the credentials operator and allows for single use execution of the reconcile loop. Provising CLI tooling allows users to bypass the security concerns of storing powerful cloud credentials in the cluster and instead keep them entirely locally.
+
+### Upgradable False Condition
+
+A field will be added to the cloud credential operator ConfigMap (possibly becoming a separate API type in another enhancement) storing a list of Updates (from ClusterVersion.AvailableUpdates) that have been reconciled, confirming that we have minted credentials and it is safe to upgrade to this release.
+
+```yaml
+  reconciledAvailableUpdates:
+  - version: 4.4.1
+    image: quay.io/openshift-release-dev/ocp-release:4.4.1-x86_64
+  - version: 4.4.2
+    image: quay.io/openshift-release-dev/ocp-release:4.4.2-x86_64
+```
+
+Entries will be added to this list by the CLI tooling below, however an administrator could do this by hand if desired.
+
+If the CCO is enabled per normal best practice install, this field will be ignored, and should be absent.
+
+However if CCO is disabled, it will compare the ClusterVersion.Spec.AvailableUpdates to this list of ReconciledAvailableUpdates. If *any* available update is not present in the reconciled list, the CCO will report Upgradable=False.
+
+### Initial Installation
+
+```bash
+$ oc adm credentials reconcile --manifests=/path/to/installer/manifests
+```
+
+This command would parse the CredentialsRequests in the manifests directory, mint credentials in the cloud using the local users environment variables or credentials file, store the status on the CredentialsRequsts (which are yet to be created), write Secret manifests containing the cloud credentials, and disable the CCO.
+
+### Upgrade Reconciliation
+
+```bash
+$ oc adm credentials reconcile --available-updates
+```
+
+This command will iterate all currently available updates in ClusterVersion.AvailableUpdates. For each given release image we will extract each CredentialsRequest and use the local users environment variables/credentials file to compare and adjust permissions as necessary.
+
+This command reconciles permissions *in an additive fashion*. We do not want to reconcile any reduction in permissions as that would affect the running components in the cluster right now. The command would print a warning if this situation is encountered, and it can be resolved after upgrade with the next command.
```

Comment:
The current CredentialsRequests are also available, so you should be able to say any of:

a. Current credential has perfect privilege for both the current release and all available updates.
b. Current credential has more privilege than required for both the current release and all available updates.
c. Current credential has sufficient privilege for the current release, but not enough for $THESE_UPDATES.
d. Current credential has insufficient privilege for the current release should already be throwing alerts, but if we don't know off the top of our heads, someone should be double-checking.

Reply:
This feels tricky. How do you determine whether going from `ec2:DescribeInstances` to `ec2:Describe*` is additive (or not) without cloud-specific parsing of permissions?
Do we just build up a user/credentials that blindly has both permissions since that should cover version CURRENT and version NEXT?

Reply:
I would expect we treat each permission as it's own entity, if it's listed, we make sure you have it, and we'll scale back down if you reconcile again post-upgrade.